src/marqo/s2_inference/clip_utils.py

```diff
@@ -475,134 +483,235 @@ def encode(self, inputs: Union[str, ImageType, List[Union[str, ImageType]]],
-        super().__init__(model_type, device,  embedding_dim, truncate , **kwargs)
+class OPEN_CLIP(AbstractCLIPModel):
+    def __init__(
+            self,
+            model_type: str = "open_clip/ViT-B-32-quickgelu/laion400m_e32",
+            device: Optional[str] = None,
+            embedding_dim: Optional[int] = None,
+            truncate: bool = True,
+            model_properties: Optional[Dict] = None,
+            model_auth: Optional[Dict] = None
+    ) -> None:
+
+        super().__init__(model_type, device, embedding_dim, truncate, model_properties)
+
+        # model_auth gets passed through add_docs and search requests:
+        self.model_auth = model_auth
+        self.preprocess_config = None
         self.model_name = model_type.split("/", 3)[1] if model_type.startswith("open_clip/") else model_type
         self.pretrained = model_type.split("/", 3)[2] if model_type.startswith("open_clip/") else model_type
+        self.model_properties = OpenCLIPModelProperties(**self.model_properties)
 
     def load(self) -> None:
-        # https://github.com/mlfoundations/open_clip
-        path = self.model_properties.get("localpath", None) or self.model_properties.get("url", None)
+        """Load the open_clip model and tokenizer."""
+        if self.model_properties.url or self.model_properties.model_location:
+            self.model, self.preprocess = self._load_model_and_image_preprocessor_from_checkpoint()
+            self.tokenizer = self._load_tokenizer_from_checkpoint()
+        elif self.model_properties.name.startswith(HF_HUB_PREFIX):
+            self.model, self.preprocess = self._load_model_and_image_preprocessor_from_hf_repo()
+            self.tokenizer = self._load_tokenizer_from_hf_repo()
+        elif self.model_tag.startswith(MARQO_OPEN_CLIP_REGISTRY_PREFIX):
+            self.model, self.preprocess = self._load_model_and_image_preprocessor_from_open_clip_repo()
+            self.tokenizer = self._load_tokenizer_from_open_clip_repo()
+        else:
+            raise InvalidModelPropertiesError(
+                f"Marqo cannot load the provided open_clip model. "
+                f"Check {marqo_docs.bring_your_own_model()} "
+                f"for more details on the supported methods to open_clip model "
+            )
+        self.model.eval()
 
-        model_location_presence = ModelProperties.model_location in self.model_properties
-        if path is None and not model_location_presence:
-            self.model, _, self.preprocess = open_clip.create_model_and_transforms(self.model_name,
-                                                                                   pretrained=self.pretrained,
-                                                                                   device=self.device, jit=False, cache_dir=ModelCache.clip_cache_path)
-            self.tokenizer = open_clip.get_tokenizer(self.model_name)
-            self.model.eval()
+    def load_tokenizer(self):
+        pass
```

Comment:
?

Reply:
I restructure the abstract model. Now the load method will call self._load_necessary_components and self._check_loaded_components.